browser mode + `resetModules()`| TypeError: Cannot read properties of undefined (reading 'idToModuleMap')

### Describe the bug

using browser mode with `resetModules()` results in a type error

```
Error: Failed to import test file /tmp/vitest-browser/vitest.setup.js
Caused by: TypeError: Cannot read properties of undefined (reading 'idToModuleMap')
```

### Reproduction

https://github.com/OrbisK/vitest-browser-mode-reproduction

1. `pnpm i`
2. `pnpm test`

### System Info

```shell
System:
    OS: Linux 6.8 Ubuntu 24.04.2 LTS 24.04.2 LTS (Noble Numbat)
    CPU: (8) x64 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
    Memory: 43.87 GB / 62.58 GB
    Container: Yes
    Shell: 5.9 - /usr/bin/zsh
  Binaries:
    Node: 23.5.0 - /home/robin/.nvm/current/bin/node
    npm: 10.9.2 - /home/robin/.nvm/current/bin/npm
    pnpm: 10.19.0 - /home/robin/.nvm/current/bin/pnpm
    bun: 1.1.42 - /home/robin/.nvm/current/bin/bun
  Browsers:
    Chrome: 130.0.6723.69
    Firefox: 144.0.2
    Firefox Developer Edition: 144.0.2
  npmPackages:
    @vitest/browser-playwright: ^4.0.5 => 4.0.5 
    vitest: ^4.0.5 => 4.0.5
```

### Used Package Manager

npm

### Validations

- [x] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [x] Read the [docs](https://vitest.dev/guide/).
- [x] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [x] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [x] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/browser/src/client/public/esm-client-injector.js
+++ b/packages/browser/src/client/public/esm-client-injector.js
@@ -1,26 +1,37 @@
 (() => {
-  const moduleCache = new Map();
-
-  function wrapModule(module) {
-    if (typeof module === "function") {
-      const promise = new Promise((resolve, reject) => {
-        if (typeof __vitest_mocker__ === "undefined")
-          return module().then(resolve, reject);
-        __vitest_mocker__.prepare().finally(() => {
-          module().then(resolve, reject);
-        });
-      });
-      moduleCache.set(promise, { promise, evaluated: false });
-      return promise.finally(() => moduleCache.delete(promise));
+  function wrapModule(moduleCallback) {
+    if (typeof moduleCallback !== "function") {
+      return moduleCallback
     }
-    return module;
+
+    if (typeof __vitest_mocker__ === "undefined" || typeof __vitest_worker__ === 'undefined') {
+      return moduleCallback()
+    }
+
+    const { evaluatedModules } = __vitest_worker__
+    const moduleId = crypto.randomUUID()
+    const viteModule = evaluatedModules.ensureModule(moduleId, moduleId)
+
+    viteModule.evaluated = false
+    viteModule.promise = new Promise((resolve, reject) => {
+      __vitest_mocker__.prepare().finally(() => {
+        moduleCallback().then(resolve, reject)
+      });
+    });
+    return viteModule.promise.finally(() => {
+      viteModule.evaluated = true
+      viteModule.promise = undefined
+
+      evaluatedModules.idToModuleMap.delete(viteModule.id)
+      evaluatedModules.fileToModulesMap.delete(viteModule.file)
+      evaluatedModules.urlToIdModuleMap.delete(viteModule.url)
+    });
   }
 
   window.__vitest_browser_runner__ = {
     wrapModule,
     wrapDynamicImport: wrapModule,
     disposeExceptionTracker: () => {},
-    moduleCache,
     cleanups: [],
     config: { __VITEST_CONFIG__ },
     viteConfig: { __VITEST_VITE_CONFIG__ },

--- a/packages/browser/src/client/tester/state.ts
+++ b/packages/browser/src/client/tester/state.ts
@@ -1,5 +1,6 @@
 import type { BrowserRPC } from '@vitest/browser/client'
 import type { WorkerGlobalState } from 'vitest'
+import { EvaluatedModules } from 'vite/module-runner'
 import { getBrowserState } from '../utils'
 
 const config = getBrowserState().config
@@ -30,8 +31,8 @@ const state: WorkerGlobalState = {
     },
   },
   onCleanup: fn => getBrowserState().cleanups.push(fn),
-  evaluatedModules: getBrowserState().evaluatedModules,
-  resolvingModules: getBrowserState().resolvingModules,
+  evaluatedModules: new EvaluatedModules(),
+  resolvingModules: new Set(),
   moduleExecutionInfo: new Map(),
   metaEnv: null as any,
   rpc: null as any,

--- a/packages/browser/src/client/utils.ts
+++ b/packages/browser/src/client/utils.ts
@@ -1,5 +1,5 @@
 import type { VitestRunner } from '@vitest/runner'
-import type { EvaluatedModules, SerializedConfig, WorkerGlobalState } from 'vitest'
+import type { SerializedConfig, WorkerGlobalState } from 'vitest'
 import type { IframeOrchestrator } from './orchestrator'
 import type { CommandsManager } from './tester/tester-utils'
 
@@ -65,8 +65,6 @@ export function ensureAwaited<T>(promise: (error?: Error) => Promise<T>): Promis
 export interface BrowserRunnerState {
   files: string[]
   runningFiles: string[]
-  resolvingModules: Set<string>
-  evaluatedModules: EvaluatedModules
   config: SerializedConfig
   provider: string
   runner: VitestRunner

--- a/test/browser/src/dynamic.ts
+++ b/test/browser/src/dynamic.ts
@@ -1,0 +1,3 @@
+window.dispatchEvent(new CustomEvent('module:loaded'))
+
+export {}

--- a/test/browser/test/dynamic-module.test.ts
+++ b/test/browser/test/dynamic-module.test.ts
@@ -1,0 +1,9 @@
+import { expect, test, vi } from 'vitest'
+
+test('can await the end of the dynamic import', async () => {
+  const _promise = import('../src/dynamic')
+  const spy = vi.fn()
+  window.addEventListener('module:loaded', spy)
+  await vi.dynamicImportSettled()
+  expect(spy).toHaveBeenCalled()
+})

